pkg/zsio/json/parser.go

```diff
@@ -0,0 +1,145 @@
+package json
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strconv"
+
+	"github.com/buger/jsonparser"
+	"github.com/mccanne/zq/pkg/zeek"
+	"github.com/mccanne/zq/pkg/zson"
+	"github.com/mccanne/zq/pkg/zval"
+)
+
+// NewRawFromJSON returns a new zson.Raw slice as well as an inferred zeek.Type
+// from provided block of JSON. The function expects the input json to be an
+// object, otherwise an error is returned.
+func NewRawFromJSON(b []byte) (zson.Raw, zeek.Type, error) {
+	val, typ, _, err := jsonparser.Get(b)
+	if err != nil {
+		return nil, nil, err
+	}
+	if typ != jsonparser.Object {
+		return nil, nil, fmt.Errorf("expected json type to be Object got %v", typ)
+	}
+	values, ztyp, err := jsonParseObject(val)
+	if err != nil {
+		return nil, nil, err
+	}
+	var raw zson.Raw
+	for _, v := range values {
+		raw = zval.AppendValue(raw, v)
+	}
+	return raw, ztyp, nil
+}
+
+func jsonParseObject(b []byte) ([][]byte, zeek.Type, error) {
+	type kv struct{ column string }
+	var columns []zeek.Column
+	var values [][]byte
+	err := jsonparser.ObjectEach(b, func(key []byte, value []byte, typ jsonparser.ValueType, offset int) error {
+		val, vtyp, err := jsonParseValue(value, typ)
+		if err != nil {
+			return err
+		}
+		values = append(values, val)
+		columns = append(columns, zeek.Column{Name: string(key), Type: vtyp})
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	// sort fields lexigraphically ensuring maps with the same
+	// columns but different printed order get assigned the same descriptor.
+	sort.Slice(columns, func(i, j int) bool {
+		v := columns[i].Name < columns[j].Name
+		if v {
+			values[i], values[j] = values[j], values[i]
+		}
+		return v
+	})
+	return values, &zeek.TypeRecord{Columns: columns}, nil
+}
+
+func jsonParseValue(raw []byte, typ jsonparser.ValueType) ([]byte, zeek.Type, error) {
+	switch typ {
+	default:
```

Comment:
Nit: I think it'd be nice to be consistent about where we put `default`, and with one exception in this repository, it follows the final `case`.